docs/architecture/adr-054-semver-compatible-modules.md

```diff
@@ -0,0 +1,493 @@
+machine in go module `example.com/foo`:
+
+```protobuf
+package foo.v1;
+
+message MsgDoSomething {
+  string sender = 1;
+  uint64 amount = 2;
+}
+
+service Msg {
+  DoSomething(MsgDoSomething) returns (MsgDoSomethingResponse);
+}
+```
+
+Now consider that we make a revision to this module and add a new `condition` field to `MsgDoSomething` and also
+add a new validation rule on `amount` requiring it to be non-zero, and that following go semantic versioning we
+release the next state machine version of `foo` as `example.com/foo/v2`.
+
+```protobuf
+// Revision 1
+package foo.v1;
+
+message MsgDoSomething {
+  string sender = 1;
+  
+  // amount must be a non-zero integer.
+  uint64 amount = 2;
+  
+  // condition is an optional condition on doing the thing.
+  //
+  // Since: Revision 1
+  Condition condition = 3;
+}
+```
+
+Approaching this naively, we would generate the protobuf types for the initial
+version of `foo` in `example.com/foo/types` and we would generate the protobuf
+types for the second version in `example.com/foo/v2/types`.
+
+Now let's say we have a module `bar` which talks to `foo` using this keeper
+interface which `foo` provides:
+```go
+type FooKeeper interface {
+	DoSomething(MsgDoSomething) error
+}
+```
+
+#### Scenario A: Backward Compatibility: Newer Foo, Older Bar
+
+Imagine we have a chain which uses both `foo` and `bar` and wants to upgrade to
+`foo/v2`, but the `bar` module has not upgraded to `foo/v2`.
+
+In this case, the chain will not be able to upgrade to `foo/v2` until `bar` 
+has upgraded its references to `example.com/foo/types.MsgDoSomething` to 
+`example.com/foo/v2/types.MsgDoSomething`.
+
+Even if `bar`'s usage of `MsgDoSomething` has not changed at all, the upgrade
+will be impossible without this change because `example.com/foo/types.MsgDoSomething`
+and `example.com/foo/v2/types.MsgDoSomething` are fundamentally different
+incompatible structs in the go type system.
+
+#### Scenario B: Forward Compatibility: Older Foo, Newer Bar
+
+Now let's consider the reverse scenario, where `bar` upgrades to `foo/v2`
+by changing the `MsgDoSomething` reference to `example.com/foo/v2/types.MsgDoSomething`
+and releases that as `bar/v2` with some other changes that a chain wants.
+The chain, however, has decided that it thinks the changes in `foo/v2` are too
+risky and that it'd prefer to stay on the initial version of `foo`.
+
+In this scenario, it is impossible to upgrade to `bar/v2` without upgrading
+to `foo/v2` even if `bar/v2` would have worked 100% fine with `foo` other
+than changing the import path to `MsgDoSomething` (meaning that `bar/v2`
+doesn't actually use any new features of `foo/v2`).
+
+Now because of the way go semantic import versioning works, we are locked
+into either using `foo` and `bar` OR `foo/v2` and `bar/v2`. We cannot have
+`foo` + `bar/v2` OR `foo/v2` + `bar`. The go type system doesn't allow this
+even if both versions of these modules are otherwise compatible with each
+other.
```

Comment:
Actually, I think go type system addresses this issue.
1. if you are using proper interface, defined in module foo, then since `MsgDoSomething` in v2 changed, you will have type error to use `foo/v2.FooKeeper` with `foo/v1.MsgDoSomething`
2. otherwise, if `FooKeeper` interface is defined in module `bar` and want's to be used as `foo/v1.Keeper` then there linking error. We want be able to pass `foo/v2.Keeper` as instance of `bar/v1.FooKeeper` because `MsgDoSomeithing` is not compatible.
3. Finally, if we think in terms protobuf bytes, then technically `v1/MsgDoSomeithing` can be deserialized as `v1/MsgDoSomeithing` (if `condition` is nullable). But we have a logic errors - we want to use API in an incompatible way. Hence probably we should not use proto version in go modules, but always keep updating module version if there is any backward incompatible change (logic or API). And always updating version of _api_ go modules.

Reply:
What do you mean by the go type system addresses this? What you're saying in 1 and 2 is I think what the ADR is trying to say - that you'll have a compile time type error.

> Finally, if we think in terms protobuf bytes, then technically v1/MsgDoSomeithing can be deserialized as v1/MsgDoSomeithing (if condition is nullable).

It doesn't matter if condition is nullable, it just needs to be left unset. Protobuf will just ignore unknown fields that are set unless we explicitly restrict that (which we do).

> Hence probably we should not use proto version in go modules, but always keep updating module version if there is any backward incompatible change (logic or API). And always updating version of api go modules.

Not sure what you mean here...